File: 207-course-schedule/207-course-schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    let dependencyMat = {};
    let visitedCourse = [];
    
    
    for(let [c1, c2] of prerequisites) {
        if(!dependencyMat[c2]) {
            dependencyMat[c2] = [c1]
        } else {
            dependencyMat[c2].push(c1);
        }
    }
    
    function isCourseVisited(node) {
        if(visitedCourse[node]){
            return false;
        }
        if(dependencyMat[node] !==undefined){
            if (dependencyMat[node].length === 0){
                return true;
            }
            visitedCourse[node] = true;
            for(let val of dependencyMat[node]){
                if(!isCourseVisited(val)){
                    return false
                }
            }
            visitedCourse[node] = false;
            dependencyMat[node] = [];
        }
        return true;

    }
    
    
    for(let k in dependencyMat) {
        if(!isCourseVisited(k)) {
            return false;
        }
    }
    return true;
    

};

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        in_degree = [0]*numCourses
        graph = {}

        for [i, j] in prerequisites:
          if i not in graph:
            graph[i] = []

          graph[i].append(j);
          in_degree[j] += 1

        queue = [i for i, degree in enumerate(in_degree) if degree == 0]

        count = 0
        while (len(queue)):
          current = queue.pop(0)
          count += 1

          for next in graph.get(current, []):
            in_degree[next] -= 1
            if in_degree[next] == 0:
              queue.append(next)

        return count == numCourses 